### ISSUE TITLE
strange eval behavior 

### ISSUE THREAD

**Issue by anch0vy** [angr/claripy#198]
## env
Ubuntu 18.04.4 LTS
Python 3.8.1
claripy==9.0.5450

## code
```python
import claripy
solver = claripy.Solver()
v = claripy.BVS("what", 64)
print(solver.eval(v * v % 2 == 0, 10))
print(solver.eval(v * v % 2 == 0, 10))
```

## expected result
(False, True)
(False, True)

## actual result
(False, True)
(True,)

**Comment by rhelmot** [comment]
This problem is a collision between the implementation of the `_eval_exhausted` cache and this behavior in the model cache mixin:

```
        # Z3 might give us solutions for variables that we did not ask for. so we create a new dict with solutions for
        # only the variables that are under the solver's control
```

Since we are querying for variables which are not part of the constraint set, this causes them to be excluded from the cached model, so all the generated models appear to be identical, and so the next time the solver is consulted, there appears to be a cached model and the expression is cached as having been exhausted, so it assumes that this model is the only model and does not consult z3 further.

I don't really know what the solution is. Maybe only adding things to the exhausted cache if their variables are all part of the solver?

**Comment by ltfish** [comment]
Assigning this issue to @zardus in case he has something to say.

**Comment by ltfish** [comment]
This has been fixed by PR #310!
